Add QEI support for LPTIM

jamwaffles opened the pull request:
As per the title, adds support for turning on encoder mode for LPTIM. e.g.

```rust
let mut pwr = PWR::new(dp.PWR, &mut rcc);
    
let lptim = LpTimer::init_encoder(
    dp.LPTIM,
    &mut pwr,
    &mut rcc,
    ClockSrc::Hsi16,
    (gpiob.pb5, gpiob.pb7),
    // Roll over at 4096
    4096,
);

defmt::info!("Count {:?}", lptim.count());
```

jamwaffles commented:
This was tested on an L053 Nucleo with a lil hand encoder module from Aliexpress.

jamwaffles commented:
I just realised it's not possible to enable interrupts with this API. Converting to draft until that's fixed...

jamwaffles commented:
The API looks like this now:

```rust
let mut encoder = LpTimer::init_encoder(
    LPTIM,
    &mut pwr,
    &mut rcc,
    // Use low speed internal oscillator
    lptim::ClockSrc::Lsi,
    (gpiob.pb5, gpiob.pb7),
);

encoder.enable_interrupts(lptim::Interrupts {
    enc_dir_up: true,
    ..lptim::Interrupts::default()
});

// Set ARR to 4096
encoder.enable(4096);
```

I don't think it's very ergonomic, but at least it works... it also reuses as much of the existing API as possible.

Very open to suggestions on how to make this nicer to use!

dbrgn commented on src/lptim.rs:
```diff
@@ -105,6 +111,81 @@ impl LpTimer<OneShot> {
+        self.lptim.arr.write(|w| w.arr().bits(arr));
+    }
+
+    pub fn enable(&mut self, arr: u16) {
```
Is there a way to disable the timer again?

dbrgn replied:
To me it's not obvious what `arr` contains, but that may be because I'm not very familiar with QEI. However, since it's a public method, it probably wouldn't hurt to add a documentation string to this function with a quick explanation.

dbrgn reviewed (commented):
I'm not really familiar with QEI (haven't used it myself so far), but I assume since @jamwaffles tested this, we can merge the PR once the comments are addressed. The code looks clean.

Maybe some other maintainers have experience with QEI?

hannobraun reviewed (approved):
I'm not familiar with QEI either, but I can't see anything wrong here. It works for @jamwaffles, so let's merge it!

jamwaffles commented:
Review comments addressed - I added a more descriptive comment for `arr`, although I left it vague as I'm unsure what it wraps around to. The value in `CMP` maybe?

I also added a `disable` method as suggested.

hannobraun reviewed (approved):
Thank you, @jamwaffles, looks good to me. Not merging, to give @dbrgn a chance to take another look.

threema-danilo commented:
Perfect, thanks @jamwaffles!

dbrgn reviewed (approved):
Perfect, thanks @jamwaffles!

jamwaffles commented:
Thanks! One less patch to keep in my fork :P
